packages/amplify-headless-interface/scripts/generateSchemas.ts

```diff
@@ -0,0 +1,49 @@
+import { getProgramFromFiles, buildGenerator, PartialArgs, Definition } from 'typescript-json-schema';
+import fs from 'fs-extra';
+import path from 'path';
+
+// paths are relative to the package root
+const typesSourceRoot = './src/interface/';
+const schemaFilesRoot = './schemas';
+
+// defines the type names and the paths to the TS files that define them
+const typeDefs: TypeDef[] = [
+  {
+    typeName: 'AddAuthRequest',
+    relativeSourcePaths: ['addAuth.ts'],
+  },
+  {
+    typeName: 'AddApiRequest',
+    relativeSourcePaths: ['addApi.ts'],
+  },
+  {
+    typeName: 'AddStorageRequest',
+    relativeSourcePaths: ['addStorage.ts'],
+  },
+];
+
+const schemaFileName = (typeName: string) => `${typeName}.schema.json`;
+const forceFlag = '--overwrite';
+const force = process.argv.includes(forceFlag);
+
+// schema generation settings. see https://www.npmjs.com/package/typescript-json-schema#command-line
+const settings: PartialArgs = {
+  required: true,
+};
+
+typeDefs.forEach(typeDef => {
+  const files = typeDef.relativeSourcePaths.map(p => path.resolve(path.join(typesSourceRoot, p)));
+  const typeSchema = buildGenerator(getProgramFromFiles(files), settings)!.getSchemaForSymbol(typeDef.typeName);
+  const version = (typeSchema.properties!.version as Definition).enum![0] as number;
+  const schemaFilePath = path.resolve(path.join(schemaFilesRoot, version.toString(), schemaFileName(typeDef.typeName)));
+  if (!force && fs.existsSync(schemaFilePath)) {
+    console.error(
+      `Schema version ${version} already exists for type ${typeDef.typeName}.\nThe interface version must be bumped after any changes.\nUse the ${forceFlag} flag to overwrite existing versions`,
+    );
+    process.exit(1);
+  }
+  fs.ensureFileSync(schemaFilePath);
+  fs.writeFileSync(schemaFilePath, JSON.stringify(typeSchema, undefined, 2));
```

Comment:
I believe we're using indentation of 4 and not 2 for our all our JSON stringification. At this point I think we should have a function in core to do this so that it's standardized across all the JSON file creations.

Reply:
Since this package is just an interface package, it shouldn't declare a dependency on the core. I'm fine with changing this indentation to 4 but my thought was that 2 would be more readable as JSON schemas can be highly nested

Reply:
Yeah, let's change it to 4 since that's the standard we use across all JSON files we generate. Long term this should be in the core shared across all the plugins to maintain standardization.

Reply:
done